Primitive type

MingzhiYe16 opened the pull request:
## Checklist
- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

MingzhiYe16 committed:
Merge branch 'datahub-project:master' into thrift-ingestion

Commit:
first commit, setup.py revised

Commit:
Add codes in the build.gradle to generate dist, which enable thrift.py to parse thrift file

Commit:
revise thrift.py

Commit:
revise build.gradle

Commit:
Merge branch 'master' of github.com:wizardxz/datahub into thrift

Commit:
revise build.gradle thrift.py test_thrift.py

Commit:
revise build.gradle

Commit:
revise antlr install process and thrift.py to adjust to MCP

Commit:
change thrift.g4 name

Commit:
revert unnecessary change

wizardxz commented on metadata-ingestion/src/datahub/ingestion/source/thrift.py:
```diff
@@ -0,0 +1,141 @@
+    ) -> SchemaFieldDataType:
+        if ctx.TYPE_I32() or ctx.TYPE_I64() or ctx.TYPE_DOUBLE():
+            return SchemaFieldDataType(NumberType())
+        elif ctx.TYPE_STRING:
```
TYPE_STRING()

Please add unit tests which cover all primitive_types

rslanka commented on metadata-ingestion/build.gradle:
```diff
@@ -66,15 +72,23 @@ task lint(type: Exec, dependsOn: installDev) {
+  classpath = buildscript.configurations.classpath
+  mainClass = 'org.antlr.v4.Tool'
+  ignoreExitValue true
+  workingDir="src/datahub/ingestion/source"
```
can we use a temporary directory instead, that is deleted after ingestion is complete?

MingzhiYe16 replied:
Sorry, I can't understand this, can you give me an example?

MingzhiYe16 replied:
we only need to generate once and it can be used for all input thrift files

rslanka replied:
You could use File.createTempDir and also specify that it be deleted on exit. Here is an example of how to do it.
https://github.com/sberbankjavaschool/homework/blob/11db332ee36bfa1f686842318a8ee68a7dd18c04/build.gradle#L77

rslanka replied:
If this task is building the lexer/parser etc thrift tools, we can even create the temporary directory under the project's buildDir.

rslanka commented on metadata-ingestion/build.gradle:
```diff
@@ -66,15 +72,23 @@ task lint(type: Exec, dependsOn: installDev) {
-    "isort src/ tests/ examples/ && " +
-    "flake8 src/ tests/ examples/ && " +
-    "mypy src/ tests/ examples/"
+    "black src/ tests/ examples/ --exclude src/datahub/ingestion/source/dist/ && " +
```
Can we eliminate this?

MingzhiYe16 replied:
No, because we don't want to lint the generated files

rslanka replied:
What I meant is that we generate the files outside in a temporary directory, not under src/datahub/ingestion/source/dist/.

MingzhiYe16 replied:
OK, I'm working on this

rslanka commented on metadata-ingestion/examples/recipes/thrift_to_console.yml:
```diff
@@ -0,0 +1,10 @@
+source:
+  type: "thrift"
+  config:
+    filename: "./examples/thrift_files/example.thrift"
```
Instead of a single file per recipe, can we accept a list of them?

MingzhiYe16 replied:
This requirement will be resolved in following diffs.

rslanka commented on metadata-ingestion/src/datahub/ingestion/source/thrift.py:
```diff
@@ -0,0 +1,141 @@
+
+
+class ThriftSourceConfig(ConfigModel):
+    filename: str
```
make this list instead?

MingzhiYe16 replied:
This requirement will be resolved in following diffs.

rslanka commented on metadata-ingestion/src/datahub/ingestion/source/thrift.py:
```diff
@@ -0,0 +1,141 @@
+)
+
+
+class ThriftSourceConfig(ConfigModel):
```
Is there anything else that we want to make configurable?

MingzhiYe16 replied:
So far, no

MingzhiYe16 replied:
TODO

rslanka replied:
TODO: add a directory option, potentially a git repo in the future.

rslanka commented on docker/elasticsearch/env/docker.env:
```diff
@@ -1 +1 @@
-ES_JAVA_OPTS="-Xms256m -Xmx256m -Dlog4j2.formatMsgNoLookups=true"
+ES_JAVA_OPTS="-Xms256m -Xmx256m -Dlog4j2.formatMsgNoLookups=true"
```
Though looks good, doesn't seem related to this PR. Can we exclude this file?

MingzhiYe16 replied:
Yes

wizardxz replied:
Can you address the comment?

rslanka commented on metadata-ingestion/src/datahub/ingestion/source/Thrift.g4:
```diff
@@ -0,0 +1,222 @@
+grammar Thrift;
```
Please create a directory called `thrift` and move both thrift.py & Thrift.g4(lower-cased name) under that.

MingzhiYe16 replied:
That's great

wizardxz replied:
Please address comments.

rslanka commented on metadata-ingestion/src/datahub/ingestion/source/Thrift.g4:
```diff
@@ -0,0 +1,222 @@
+grammar Thrift;
```
Could you add a comment on this file is defining and if there are any current limitations, including a link to the documentation?

MingzhiYe16 replied:
OK

rslanka commented on metadata-ingestion/src/datahub/ingestion/source/thrift.py:
```diff
@@ -0,0 +1,141 @@
+        if ctx.struct_():
+            yield from self.bind_MCPs_from_struct_(ctx.struct_())
+        else:
+            raise NotImplementedError()
```
Is this just defensive or are there incomplete unhandled cases? Please add a comment.

MingzhiYe16 replied:
OK

rslanka commented on metadata-ingestion/src/datahub/ingestion/source/thrift.py:
```diff
@@ -0,0 +1,141 @@
+    ) -> Generator[MetadataChangeProposalWrapper, None, None]:
+
+        name = ctx.IDENTIFIER().getText()
+        urn = f"urn:li:dataset:(urn:li:dataPlatform:thrift,{name},PROD)"
```
Please use https://github.com/datahub-project/datahub/blob/master/metadata-ingestion/src/datahub/emitter/mce_builder.py#L69 to construct the urn.

MingzhiYe16 replied:
OK

rslanka commented on metadata-ingestion/src/datahub/ingestion/source/thrift.py:
```diff
@@ -0,0 +1,141 @@
+        if ctx.base_type():
+            return ctx.getText()
+        else:
+            raise NotImplementedError()
```
Comments about unhandled cases.

MingzhiYe16 replied:
OK

rslanka commented on metadata-ingestion/src/datahub/ingestion/source/thrift.py:
```diff
@@ -0,0 +1,141 @@
+        if ctx.base_type():
+            return self.bind_SchemaFieldDataType_from_base_type(ctx.base_type())
+        else:
+            raise NotImplementedError()
```
comments about unhandled cases.

MingzhiYe16 replied:
OK

wizardxz replied:
Please address comments

rslanka commented on metadata-ingestion/src/datahub/ingestion/source/thrift.py:
```diff
@@ -0,0 +1,141 @@
+        elif ctx.TYPE_STRING:
+            return SchemaFieldDataType(StringType())
+        else:
+            raise NotImplementedError()
```
comments about unhandled types.

MingzhiYe16 replied:
OK

rslanka commented on metadata-ingestion/src/datahub/ingestion/source/thrift.py:
```diff
@@ -0,0 +1,141 @@
+@dataclass
+class ThriftSource(Source):
+    config: ThriftSourceConfig
+    report: SourceReport = field(default_factory=SourceReport)
```
Are there any interesting metrics that we want to report, such as errors and processing time?

MingzhiYe16 replied:
This requirement will be resolved in following diffs.

rslanka replied:
You can create a sub-class of SourceReport, called ThriftReport to extend it.

MingzhiYe16 replied:
Hello, I'm not very familiar with report, can you give me an example?

rslanka replied:
You can look at [this example](https://github.com/datahub-project/datahub/blob/master/metadata-ingestion/src/datahub/ingestion/source_report/sql/bigquery.py). There are more under [this](https://github.com/datahub-project/datahub/tree/master/metadata-ingestion/src/datahub/ingestion/source_report).

MingzhiYe16 replied:
I have created a sub-class ThriftReport and add variables: start time and end time. I found that failures{} has been in SourceReport, so I don't need to add errors{} to ThriftReport, yes?

rslanka replied:
Correct. We could later extend it to get other fine-grained information like total number of files & structs that have been successfully processed, any metrics around ownership/documentation/tags/terms etc that we think are a useful summary to track.

rslanka commented on metadata-ingestion/src/datahub/ingestion/source/thrift.py:
```diff
@@ -0,0 +1,141 @@
+        with open(filename) as text_file:
+            # lexer
+
+            lexer = ThriftLexer(InputStream(text_file.read()))
```
Are there any exceptions in this block that we want to catch and report?

MingzhiYe16 replied:
This requirement will be resolved in following diffs.

MingzhiYe16 replied:
Is it OK to do these here and add found errors to a metric of report: self.report.errors: List[str]?


                lexer = ThriftLexer(InputStream(text_file.read()))
                stream = CommonTokenStream(lexer)
                # parser
                parser = ThriftParser(stream)

                tree = parser.document()
                # evaluator
                yield from Binder().bind_MCPs_from_Document(tree)
        except Exception as e:
            self.report.errors.append(f"Error: {e}")`

rslanka commented on metadata-ingestion/src/datahub/ingestion/source/thrift.py:
```diff
@@ -0,0 +1,141 @@
+    def get_report(self):
+        return self.report
+
+    def close(self):
```
This is a good place to delete temporary directories.

rslanka replied:
This may no longer be needed if we specify tempDir.deleteOnExit in the gradle file.

rslanka commented on metadata-ingestion/tests/unit/test_thrift.py:
```diff
@@ -0,0 +1,33 @@
+    )
+    mcp_objects = [wu.metadata.to_obj() for wu in source.get_workunits()]
+
+    with open(str(tmp_path / "thrift_mcps.json"), "w") as f:
```
Instead of the golden files based testing, please write tests against the in-memory python objects directly. See [this](https://github.com/datahub-project/datahub/blob/master/smoke-test/tests/test_stateful_ingestion.py) for an example. Feel free to ask for more if needed.

MingzhiYe16 replied:
Hello, it sames that in the file you provided, the test is for checking whether pipline works correct, yes? In my test file, I don't use mysql pipline or any complex pipline, what should I check? I just read input file, ingest and compare with golden file.

MingzhiYe16 replied:
Does checkoutpoint(in line 98 of [this](https://github.com/datahub-project/datahub/blob/8d281fc013080e1f955107774f4c0ce9639aab81/smoke-test/tests/test_stateful_ingestion.py#L98)) contains information about whether the pipline work well and we use this to check whether pipline work well?

rslanka replied:
In your test, you would be writing unit-tests directly against the key contents of the mcp_objects, such as the fieldPath, type and nativeDataType of the schemaFields produced. [This](https://github.com/datahub-project/datahub/blob/master/metadata-ingestion/tests/unit/test_kafka_source.py) is another example of such a test(we are generally trying to move away from golden file based tests, as they have a lot of irrelevant data that can change due to logic external to this connector). If you need help with a sample, I can help update this test itself, please let me know.

MingzhiYe16 replied:
Should I still read the same example thrift file, but only check whether some critical features(check whether second field's nativetype is i32 and whether second field's type is {"com.linkedin.schema.NumberType": {}}) are right instead of compare with golden file?

rslanka replied:
Yes, that's correct. We basically don't want to use the golden files. It is perfectly okay to use as many thrift files as needed for the tests.

Commit:
revised primitive_types

wizardxz commented on metadata-ingestion/src/datahub/ingestion/source/thrift.py:
```diff
@@ -0,0 +1,492 @@
+@dataclass(frozen=True)
+class GetHeader(ThriftVisitor):
+    filename: str
+    thrift_paths: Tuple[str, ...] = field(default_factory=tuple)
```
if thrift_paths is not used in GetHeader, please remove it

wizardxz commented:
Please merge it after adding TODO

rslanka committed:
Exclude the antlr generated thrift parser from mypy type checking.

rslanka committed:
1. All outstanding linter related issues.
2. Renaming of dist to parse_tools + gradle task name update.
3. Adding in the antlr generated parse_tools to the repo.

rslanka committed:
Minor fix.

rslanka committed:
Remove dist from clean
